src/vec/FixedPairList.cpp

```diff
@@ -0,0 +1,345 @@
+    size_t const p2 = storageVec->lookupLocalParticleVec(pid2);
+
+    if(p1==VEC_PARTICLE_NOT_FOUND) {
+      // Particle does not exist here, return false
+      returnVal=false;
+    } else {
+      if(p2==VEC_PARTICLE_NOT_FOUND) {
+        LOG4ESPP_DEBUG(theLogger, "Particle p2 " << pid2 << " not found");
+      }
+    }
+
+    if(returnVal){
+      // add the pair locally
+      this->push_back({p1,p2});
+
+      // ADD THE GLOBAL PAIR
+      // see whether the particle already has pairs
+      std::pair<GlobalPairs::const_iterator,
+        GlobalPairs::const_iterator> equalRange
+        = globalPairs.equal_range(pid1);
+      if (equalRange.first == globalPairs.end()) {
+        // if it hasn't, insert the new pair
+        globalPairs.insert(std::make_pair(pid1, pid2));
+      }
+      else {
+        // otherwise test whether the pair already exists
+        for (GlobalPairs::const_iterator it = equalRange.first; it != equalRange.second; ++it) {
+          if (it->second == pid2) {
+            // Pair already exists, generate error!
+            throw std::runtime_error("Duplicate pair inserted");
+            ;
+          }
+        }
+        // if not, insert the new pair
+        globalPairs.insert(equalRange.first, std::make_pair(pid1, pid2));
+      }
+      LOG4ESPP_INFO(theLogger, "added fixed pair to global pair list");
+    }
+    LOG4ESPP_DEBUG(theLogger, "Leaving add with returnVal " << returnVal);
+    return returnVal;
+  }
+
+  python::list FixedPairList::getBonds()
+  {
+    python::tuple bond;
+    python::list bonds;
+    for (GlobalPairs::const_iterator it=globalPairs.begin(); it != globalPairs.end(); it++) {
+      bond = python::make_tuple(it->first, it->second);
+      bonds.append(bond);
+    }
+
+    return bonds;
+  }
+
+  std::vector<size_t> FixedPairList::getPairList()
+  {
+    std::vector<size_t> ret;
+    for (GlobalPairs::const_iterator it = globalPairs.begin(); it != globalPairs.end(); it++) {
+      ret.push_back(it->first);
+      ret.push_back(it->second);
+    }
+    return ret;
+  }
+
+  python::list FixedPairList::getAllBonds()
+  {
+    std::vector<size_t> local_bonds;
+    std::vector<std::vector<size_t> > global_bonds;
+    python::list bonds;
+
+    for (GlobalPairs::const_iterator it = globalPairs.begin(); it != globalPairs.end(); it++) {
+      local_bonds.push_back(it->first);
+      local_bonds.push_back(it->second);
+    }
+    System& system = vectorization->getSystemRef();
+    if (system.comm->rank() == 0) {
+      mpi::gather(*system.comm, local_bonds, global_bonds, 0);
+      python::tuple bond;
+
+      for (std::vector<std::vector<size_t> >::iterator it = global_bonds.begin();
+           it != global_bonds.end(); ++it) {
+        for (std::vector<size_t>::iterator iit = it->begin(); iit != it->end();) {
+          size_t pid1 = *(iit++);
+          size_t pid2 = *(iit++);
+          bonds.append(python::make_tuple(pid1, pid2));
+        }
```

Comment:
I think readability can be improved by
```suggestion
        for (auto iit = it->begin(); iit != it->end(); iit += 2) {
          size_t pid1 = *(iit);
          size_t pid2 = *(iit + 1);
          bonds.append(python::make_tuple(pid1, pid2));
        }
```